### ISSUE TITLE
GML float attribute inconsistencies

### ISSUE THREAD

**Issue by RicoFio** [igraph/python-igraph#562]
**Describe the bug**
The GML format, by [definition](https://docs.yworks.com/yfiles/doc/developers-guide/gml.html), should be able to handle floating point / double edge or vertex attributes. However, when writing an igraph `Graph` object to a `gml` file, the floating point precision is capped to 16 digits after the comma.

**To reproduce**
```
import igraph as ig
import numpy as np
np.random.seed(12345)

g = ig.Graph(directed=True)

g.add_vertices([1,2])

len_1 = np.random.random()
g.add_edge(0,1, length=len_1)

g.write_gml('./test_write.gml')
g.write_picklez('./test_write.picklez')
```
When then loading the same graph from the two different files, this is the result:
```
g_gml = ig.read('./test_write.gml')
g_pklz = ig.read('./test_write.picklez')

g_gml.es['length'] == g_pklz.es['length']
>>> False
g_gml.es['length']
>>> 0.929616092817148
g_pklz.es['length']
>>> 0.9296160928171479
```

**Expected Behaviour**
There should not be a truncation. However, if this is part of the GML standard (and I am not aware of that), then I would expect a warning/note of this aspect in the documentation. When using iGraph, particularly for scientific purposes, this is a deal-breaker, when the imposed precision on attribute values is needed when the graph is reloaded from disk.

**Version information**
`igraph==0.9.10` in `python==3.9.12`

**Comment by szhorvat** [comment]
Please test with the development version from the `develop` branch (to become 0.10). Compilation/installation instructions are [here](https://github.com/igraph/python-igraph/#compiling-the-development-version). The GML reader/writer has undergone major refactoring for 0.10.

**Comment by szhorvat** [comment]
I just checked, and the behaviour is the same in the latest version. I did in fact look at this when overhauling the GML writer. It writes [`DBL_DIG`](https://en.cppreference.com/w/c/types/limits) _decimal_ digits. This is the number of decimal digits guaranteed to be preserved after a decimal -> binary -> decimal conversion round-trip.

You seem to be asking to write enough digits for a lossless binary -> decimal -> binary conversion. Unfortunately, the C99 standard does not provide any constants to determine how many digits this requires, so if we do this, it will need to be done in a non-portable way (which is error prone and generally increases code complexity).

It is standard practice to use `DBL_DIG` digits for serialization. You will find the same behaviour in most other systems. You say that this minor precision loss is a "deal-breaker" for scientific applications. It seems to me that this is a major overstatement. Doing this loses at most one binary digits of precision. That means that 52 binary digits of precision are guarantees instead of 53. Can you give an example of an application where this is truly a deal-breaker?

I would be willing to look into improvements here, but only if I see a convincing practical need. Can you demonstrate one?

**Comment by szhorvat** [comment]
P.S. Regarding "the GML standard", there isn't one. What you linked to describes yEd's handling of GML files. The original description of the GML format by its authors is long gone, and only accessible through archive.org, https://web.archive.org/web/20190207140002/http://www.fim.uni-passau.de/index.php?id=17297%26L=1  So there is little point in talking about complying with the standard. Instead, igraph tries to be a good citizen and interoperate with other software to a reasonable extent without introducing major format extensions.

However, the main point is: With just about _any_ textual format (not just GML), do not expect to get more precision than this. Even if igraph were changed to handle GML differently, many other software won't. If you truly need lossless storage of floating-point numbers, use a binary format such as HDF5.

**Comment by RicoFio** [comment]
Ha, you were quicker! Thanks for checking.

Okay, I see where this comes from now and I would definitely agree with you that changing this to a non-compatible format would be counter-productive. However, I would still put a warning in the documentation because it was quite disheartening when I discovered that results differed between in-memory graph and GML-file loaded graphs _on the day of my master thesis submission_ :smiling_face_with_tear:.

My use case might be very specific, but for my research I modeled public transportation network reduction (i.e. edge deletion) and its effects on public access equality. I used igraph `Graph` objects to represent the public transport and pedestrian network. 

To quantify the accessibility, I ran shortest-path routing from a set of origin to a set of destination nodes and took the summed travel time (leading to a cumulative error in this case) to compute a reward (based on the Theil inequality index, but let's not dive into this). The same rewards for the same sub-graphs differed so greatly between the in-memory graph objects and the GML-file loaded graphs, that the optimal subgraphs (the ones with maximum reward) differed in some examples.

In the example below, you can see all the possible combinations of removed edges (up to 3 out of 4) from the same graph. On the left I generate them from the in-memory `Graph` object and on the right from the GML-loaded `Graph` object. The red color in the graph's edges indicate that these have been removed from that specific graph configuration. 

Depending on the configuration, there is a reward which is computed as a function of the cumulative travel time from the red vertices to the green vertices. I have boxed in red (on the right, the GML loaded graph) the rewards which differ from the true rewards (left, in-memory graph). Furthermore, I have indicated the paths which a greedy algorithm would take, always maximizing the reward from one edge removal to the next one. The final result is the same _in this case_, however the intermediary results differ greatly and the results from the GML-loaded graph on the right are just _wrong_. I also have another example where the final optimal sub-graph is a completely different one, but this should suffice to illustrate my point.

![image](https://user-images.githubusercontent.com/3929674/185794509-9425a2b4-3efd-4970-94ff-6c3793b8d681.png)

As I said, this might be a very unique use case, but hey, I would have wished to know this limitation 4 weeks earlier and not re-run all my experiments on submission day when I discovered this while doing my final read-over XD.

**Comment by RicoFio** [comment]
> However, the main point is: With just about _any_ textual format (not just GML), do not expect to get more precision than this. Even if igraph were changed to handle GML differently, many other software won't. If you truly need lossless storage of floating-point numbers, use a binary format such as HDF5.

Yeah, I figured that out by now. I'm using `picklez` now but I might switch to HDF5.

**Comment by szhorvat** [comment]
> intermediary results differ greatly and the results from the GML-loaded graph on the right are just _wrong_.


Changing edge lengths by a relative amount on the order of $10^{-15}$ should not cause a difference in results if the problem is well formulated and the calculation is solid. You say that such a change leads to results which are wrong. Are you sure that the original results were correct?

Is it possible that your graph has several shortest paths of equal length between some pairs of vertices? For example, if one path is made of up three edges of lengths of 1/3, and the other of two of lengths 1/2, then they both have total length of 1. But if we do the calculation with 2 decimal digits of precision we get 0.33 + 0.33 + 0.33 = 0.99 < 1.0 = 0.5 + 0.5. So it would appear that there is only one shortest ("optimal") path when in fact there are two.

This same problem appears when calculating betweenness centrality, which counts the shortest paths passing through each vertex or edge. When there are multiple paths of the same length, floating point roundoff errors can cause significant differences in betweenness results. I have seen results change just from changing the optimization level of the compiler.

igraph's betweenness and `get_all_shortest_paths()` functions compensate for this by performing comparisons with a tolerance. As far as I'm aware, this is a unique feature of igraph among all network libraries.

So, if you're computing shortest paths, and a tiny change in edge lengths leads to a change in the result, check whether you have multiple shortest paths using the `get_all_shortest_paths()` function.

**Comment by szhorvat** [comment]
There is one more thing concerning this that came to mind.

Numbers that are representable in decimal are not necessarily exactly representable in binary as well. This leads to unusual effects such as `printf("%.17g\n", 100.1)` outputting `100.09999999999999` instead of `100.1`. Truncating to `DBL_DIG` digits, i.e. using `printf("%.15g\n", 100.1)` prevents this. Generally, people expect that numbers that they input in decimal do not change when they are output again. It seems that due the "incompatibility" of decimal and binary representations and the `printf` rounding/truncating behaviour, we have a choice to make: either we preserve the binary representations exactly, or we preserve the decimal ones, but we can't easily preserve both. I think that with textual formats such as GML, GraphML, Pajek, etc. it is more valuable to preserve the decimal form. In other words, writing a file with igraph, then reading it and writing it again, should produce a file identical to the one originally written. This unfortunately is in conflict with the also reasonable requirement that writing then reading back in a file should produce a data structure identical to the original. However, this second requirement can't be satisfied anyway due to the various limitations of typical graph formats. So I am in favour of preserving the textual representation losslessly.

**Comment by szhorvat** [comment]
Added a comment in the source code in case this comes up in the future:

https://github.com/igraph/igraph/blob/3e9e3dbacb9788fbb00f3035e5aaea4625b539f3/src/core/printing.c#L32
